docs/design/001-prometheus.md

```diff
@@ -0,0 +1,451 @@
+# Prometheus support in Inspektor Gadget
+
+Inspektor Gadget has a lot of tools that hook into the kernel to capture different events like file
+opened, process created, DNS requests, etc. Currently it's mostly designed as a troubleshooting
+tool: it prints those events as they happen to the terminal. However, it's an easy win to provide
+metrics through Prometheus. The whole logic to capture the data is already in place, we only need to
+aggregate and expose this in a Prometheus format.
+
+This document contains a design proposal for supporting Prometheus metrics in Inspektor Gadget.
+Upstream issue:
+[https://github.com/inspektor-gadget/inspektor-gadget/issues/1513](https://github.com/inspektor-gadget/inspektor-gadget/issues/1513)
+
+# Goals
+
+This document is written with the following goals in mind in descending order of priority.
+
+- Bring this support soon to market
+- Metrics to expose should be configurable
+- The solution should be performant
+
+# Design Decisions
+
+The following are key aspects for this support.
+
+## Metrics to Expose
+
+The user should be able to configure the metrics they want to expose for each gadget. This makes
+this support much more flexible and lets the responsibility of which metrics to expose to the user.
+One drawback of this is the performance, however in the implementation section we propose a way to
+increase performance for the most common metrics.
```

Comment:
```suggestion
In order to be as flexible as possible, the user should be able to configure the metrics they want to expose
for each gadget. Most gadgets emit events from eBPF including several fields of data and send them to the
userspace part of IG for processing. In a generic solution, most of these fields should be selectable for metric collection, aggregation and filtering. However, due to handling all events in userspace, this could negatively
impact performance. In order to improve that, we also propose a way to handle collection of the most commonly
used metrics directly in eBPF.
```